protocol/listoffsets/listoffsets.go

```diff
@@ -0,0 +1,230 @@
+func (r *Response) Reduce(requests []protocol.Message, results []interface{}) (protocol.Message, error) {
+	type topicPartition struct {
+		topic     string
+		partition int32
+	}
+
+	// Kafka doesn't always return the timestamp in the response, for example
+	// when the request sends -2 (for the first offset) it always returns -1,
+	// probably to indicate that the timestamp is unknown. This means that we
+	// can't correlate the requests and responses based on their timestamps,
+	// the primary key is the topic/partition pair.
+	//
+	// To make the API a bit friendly, we reconstructing an index of topic
+	// partitions to the timestamps that were requested, and override the
+	// timestamp value in the response.
+	timestamps := make([]map[topicPartition]int64, len(requests))
+
+	for i, m := range requests {
+		req := m.(*Request)
+		ts := make(map[topicPartition]int64, len(req.Topics))
+
+		for _, t := range req.Topics {
+			for _, p := range t.Partitions {
+				ts[topicPartition{
+					topic:     t.Topic,
+					partition: p.Partition,
+				}] = p.Timestamp
+			}
+		}
+
+		timestamps[i] = ts
+	}
+
+	topics := make(map[string][]ResponsePartition)
+	errors := 0
+
+	for i, res := range results {
+		m, err := protocol.Result(res)
+		if err != nil {
+			for _, t := range requests[i].(*Request).Topics {
+				partitions := topics[t.Topic]
+
+				for _, p := range t.Partitions {
+					partitions = append(partitions, ResponsePartition{
+						Partition:   p.Partition,
+						ErrorCode:   -1, // UNKNOWN, can we do better?
+						Timestamp:   -1,
+						Offset:      -1,
+						LeaderEpoch: -1,
+					})
+				}
+
+				topics[t.Topic] = partitions
+			}
+			errors++
+			continue
+		}
+
+		response := m.(*Response)
+
+		if r.ThrottleTimeMs < response.ThrottleTimeMs {
+			r.ThrottleTimeMs = response.ThrottleTimeMs
+		}
+
+		for _, t := range response.Topics {
+			for _, p := range t.Partitions {
+				if timestamp, ok := timestamps[i][topicPartition{
+					topic:     t.Topic,
+					partition: p.Partition,
+				}]; ok {
+					p.Timestamp = timestamp
+				}
+				topics[t.Topic] = append(topics[t.Topic], p)
+			}
+		}
+
+	}
+
+	if errors > 0 && errors == len(results) {
+		_, err := protocol.Result(results[0])
+		return nil, err
+	}
+
+	r.Topics = make([]ResponseTopic, 0, len(topics))
+
+	for topicName, partitions := range topics {
+		r.Topics = append(r.Topics, ResponseTopic{
+			Topic:      topicName,
+			Partitions: partitions,
+		})
+	}
+
+	sort.Slice(r.Topics, func(i, j int) bool {
+		return r.Topics[i].Topic < r.Topics[j].Topic
+	})
+
+	for _, t := range r.Topics {
+		sort.Slice(t.Partitions, func(i, j int) bool {
+			p1 := &t.Partitions[i]
+			p2 := &t.Partitions[j]
+
+			if p1.Partition != p2.Partition {
+				return p1.Partition < p2.Partition
+			}
+
+			return p1.Offset < p2.Offset
+		})
+	}
+
+	return r, nil
+}
+
+var (
+	_ protocol.BrokerMessage = (*Request)(nil)
+	_ protocol.Mapper        = (*Request)(nil)
+	_ protocol.Reducer       = (*Response)(nil)
```

Comment:
I'm not finding the definition of `protocol.Reducer` now, but what do you think about `Merge` instead of `Reduce`?  The map/reduce verbiage really throws me b/c it seems out of place in a kafka context.  Still thinking on what to name `Map`.  Also, technically the map operation here is more of a flat map. 😉   `FanOut`?

Reply:
`Merge` sounds good to me 👍

Reply:
How about `Dispatch` instead of `Map`?

Reply:
`Split` can be descriptive as well (split as an opposite to merge).

Reply:
`Split` sounds good!